### PR TITLE
Prime checker

### PR DESCRIPTION
**Pull request by robiel0143**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/prime_checker.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    # Raises a TypeError if the input is a String
+    if n is str:
+        raise TypeError("Input cannot be a String")
+
+    # Raises an error if the given number is negative
+    if n < 0:
+        raise ValueError("Input cannot be Negative")
+
+    # Checks if the given input is integer or Float
+    assert isinstance(n, (int, float)), "Input must be an integer or a Whole float"
```
Comment (linahKhayri) [inline solutions/prime_checker.py:52]: Can all the str and none assertions be replaced with only assert isinstance(n, (int, float)), or raising ValueError and TypeError is necessary to show an explanation why input isn't valid? just a genuine questions not a suggestion 🧐

File: solutions/tests/test_prime_checker.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    def test_number_one(self):
+        """It should return False"""
+        self.assertEqual(prime_checker(1), False)
+
+    def test_large_number(self):
+        """It should return"""
+        self.assertEqual(prime_checker(100019), True)
+
+    def test_single_digit(self):
```
Comment (linahKhayri) [inline solutions/tests/test_prime_checker.py:69]: test for single digit was previously conducted in the standard cases since the tests already involved numbers with a single digit, instead how about a more than one digit test?

Comment (robiel0143) [reply solutions/tests/test_prime_checker.py:69]: You are right but I included it with the large numbers test

File: solutions/prime_checker.py (hunk 1 above)
Comment (robiel0143) [reply solutions/prime_checker.py:52]: That is a good idea I will try it join str and None but since the comments are different I have decided to make them on a different assertion cases to give a different raise information...don't you think it is a good idea but let me try to merge them

File: solutions/tests/test_prime_checker.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        self.assertEqual(prime_checker(5), True)
+
+    def test_two_digits(self):
+        """It should return True or False based on the given number"""
+        self.assertEqual(prime_checker(23), True)
+
+    def test_whole_floats(self):
+        """It should return True"""
+        self.assertEqual(prime_checker(7.0), True)
+
```
Comment (linahKhayri) [inline solutions/tests/test_prime_checker.py:80]: How about an edge case for testing infinity?

**Review by linahKhayri — COMMENTED** [review]
You did an awesome job Robel 👏 The code checks all the boxes from the checklist, and I really like the strategy you took.✔
I just have a comment and questions that I’ve attached.

File: solutions/prime_checker.py (hunk 1 above)
Comment (linahKhayri) [reply solutions/prime_checker.py:52]: It was just a question but since there're different assertion info then I think your code is good ✔

File: solutions/prime_checker.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+Module contents:
+    - prime_checker: checks if a given number is prime or not.
+
+Created on 2024-12-31
+Author: Robel Mengsteab
+"""
+
+
+def prime_checker(n: int) -> bool:
```
Comment (Khusro-S) [inline solutions/prime_checker.py:14]: Even though `n` is clear in this context, it’s best practice to use more descriptive names. Something like `number` or a name that shows what it represents would make the code easier to understand.

File: solutions/prime_checker.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    """Checks if a number is prime.
+
+    A prime number is greater than 1 and divisible only by 1 and itself.
+
+
+    Parameters:
+        n: int, the number to greater than one
```
Comment (Khusro-S) [inline solutions/prime_checker.py:21]: Same goes here, the variable name as I mentioned above.

File: solutions/prime_checker.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+    # Convert to integer if it's a whole number float
+    n = int(n)
+
+    # Checks if the given number is less than or equal to one
+    if n <= 1:
+        return False
+
+    # Checks whether the numbers greater than one are prime or not
+    for i in range(2, int(n**0.5) + 1):
```
Comment (Khusro-S) [inline solutions/prime_checker.py:68]: Just wanted to say I really liked how you implemented this. The way you checked for primality using the square root is a smart and efficient approach.

This loop iterates through numbers `i`, starting from `2` and going up to the square root of `n`, rounded down, plus 1.
A factor of a number `n` must be less than or equal to its square root. If no factors are found in this range, `n` is `prime`.

Nicely done.

File: solutions/tests/test_prime_checker.py (hunk 3 above)
Comment (robiel0143) [reply solutions/tests/test_prime_checker.py:80]: Hey Linah, Including tests for infinity values is unnecessary for a prime checker function, because such values are not valid inputs for determining primality, also if since I had already included a function that asserts when the input must be a positive integer or a whole float, it should handle infinity naturally because float('inf') would fail the primality logic (n != int(n), so its not necessary to include it..what do you think

**Review by robiel0143 — COMMENTED** [review]
hey Linah thank you for your **_constructive helpful comments_**, I have put comments on your helpful comments so...please check them out

**Review by Khusro-S — COMMENTED** [review]
Nicely done! Just a suggestion. It's not a big issue, but for the sake of practicing best practices, it would be good to maintain descriptive naming conventions.

File: solutions/tests/test_prime_checker.py (hunk 3 above)
Comment (linahKhayri) [reply solutions/tests/test_prime_checker.py:80]: Yeah I get the point That the function doesn't need any modifications or extra code to handle this case. I just thought maybe adding a defensive test that raises an assertion for this case could be useful. But since this is a mathematical logic and your tests already cover the assertions in the function then it's good enough.
Good job 👏
